File: DataAlgo_forPy/Arrays/dynamic_array.rb
class DynamicArray
  attr_reader :count, :capacity

  def initialize
    @count = 0
    @capacity = 1
    @my_array = make_array(@capacity)
  end


  def get_item(index)
    unless 0 <= index < @count
      throw "index is out of bounds"
    end

    @my_array[index]
  end


  def append(element)
    if @capacity == @count
      self._resize(2 * @capacity)
    end

    @my_array[@count] = element
    @count += 1
  end

  def _resize(new_capacity)
    new_array = make_array(new_capacity)

    @count.times do |i|
      new_array[i] = @my_array[i]
    end

    @my_array = new_array
    @capacity = new_capacity
  end

  def make_array(new_capacity)
    Array.new(new_capacity)
  end

end


arr = DynamicArray.new
arr.append(1)
arr.append(2)
p arr.capacity
p arr.count

File: DataAlgo_forPy/Arrays/dynamic_array.js
class DynamicArray {
  constructor() {
    this.count = 0
    this.capacity = 1
    this.my_array = this.make_array(this.capacity)
  }

  get_item(index) {
    if (0 <= index && index < this.count) {
       throw "Index is out of bound"
    }
    return this.my_array[index]
  }


  append(element) {
    if (this.capacity === this.count) {
      this.resize(2 * this.capacity)
    }

    this.my_array[this.count] = element
    this.count++
  }

  resize(new_capacity) {
    const new_array = this.make_array(new_capacity)

    for (let i = this.count; i < this.capacity; i++) {
      new_array[i] = this.my_array[i]
    }

    this.my_array = new_array
    this.capacity = new_capacity
  }

  make_array(new_capacity) {
    return new Array(new_capacity)
  }

}

const arr = new DynamicArray
arr.append(1)
arr.append(2)
arr.append(3)
arr.append(3)
arr.append(3)
console.log("capacity", arr.capacity)
console.log("count", arr.count)
